core/metacling/src/TCling.cxx

```diff
@@ -3874,7 +3874,8 @@ void TCling::SetClassInfo(TClass* cl, Bool_t reload)
 
    }
 
-   TClingClassInfo* info = new TClingClassInfo(GetInterpreterImpl(), name.c_str());
+   bool instantiateTemplate = !cl->TestBit(TClass::kUnloading);
+   TClingClassInfo* info = new TClingClassInfo(GetInterpreterImpl(), name.c_str(), instantiateTemplate);
```

Comment:
Shouldn't we have a separate class for this. Eg: `TClingClassInfoReadOnly` (name is a placeholder). This would make the code more explicit that this is not changing stuff, thread-safe, etc?

Reply:
That sounds a bit heavy-handed though and would easily (not unavoidably) lead to code duplication (either way it is quite a bit more of typing).

Reply:
Could you just then add a `FIXME: We should move this functionality in a separate class TClingClassInfoReadOnly which makes the intent more explicit.`?

Reply:
Well in addition to moving that functionality the new code would need to suspend the autoparsing and autoloading (this code path by itself only disable template instantiation).
I will add a fixme nonetheless.